src/librustc_codegen_llvm/intrinsic.rs

```diff
@@ -146,6 +147,88 @@ impl IntrinsicCallMethods<'tcx> for Builder<'a, 'll, 'tcx> {
                 let tp_ty = substs.type_at(0);
                 self.cx().const_usize(self.cx().size_of(tp_ty).bytes())
             }
+            func @ "va_start" | func @ "va_end" => {
+                let va_list = if (tcx.sess.target.target.arch == "aarch64" ||
+                                  tcx.sess.target.target.arch == "x86_64" ||
+                                  tcx.sess.target.target.arch == "powerpc") &&
+                                 !tcx.sess.target.target.options.is_like_windows {
+                    self.load(args[0].immediate(), tcx.data_layout.pointer_align.abi)
+                } else {
+                    args[0].immediate()
+                };
+                let intrinsic = self.cx().get_intrinsic(&format!("llvm.{}", func));
+                self.call(intrinsic, &[va_list], None)
+            }
+            "va_copy" => {
+                if (tcx.sess.target.target.arch == "aarch64" ||
+                                     tcx.sess.target.target.arch == "x86_64" ||
+                                     tcx.sess.target.target.arch == "powerpc") &&
+                                    !tcx.sess.target.target.options.is_like_windows {
```

Comment:
You can probably just check here whether the return type(`result.layout.ty`) is `tcx.lang_items().va_list()` or something like that.

Reply:
Yup... also much cleaner